feat: add migrate project from Renku V1 to Renku 2.0

andre-code opened the pull request:
This PR introduces the ability for project owners to migrate their projects from Renku v1 to Renku 2.0. 

**Acceptance Criteria**

- The migration alert is shown only to project owners and for public projects.

- The migration process will transfer:
  1. Code repository
  2. Container image (pinned images or the last built image)
  3. Resource class (based on project specifications or a default resource class if available)
  4. Project description
  5. Project keywords

- The migration process will display what will not be migrated.
- The project is migrated, and a new launcher is added in Renku 2.0.
- Sessions should work if they worked in Renku v1 and used a public image.
- If a project requires activation in Renku v1 for a session to work, migration is blocked until the project is updated.
- Users can navigate from the original project in Renku v1 to the corresponding migrated project in Renku 2.0.



**Migrating a project...**

![migrate metadata and launcher](https://github.com/user-attachments/assets/adfcfcb5-85d8-4bba-a996-9ab1537ea693)



**Launching a session migrated project:**
![launch migrated session](https://github.com/user-attachments/assets/aefb8be4-e44f-42a3-8934-838c682dd25c)


/deploy renku=release-0.66.0 renku-data-services=andrea/add-projects-migrated renku-notebooks=master

leafty commented on client/src/features/projects/projectMigration.api.ts:
```diff
@@ -0,0 +1,21 @@
+
+export const projectMigrationApi = createApi({
+  reducerPath: "renku_v1_projects",
+  baseQuery: fetchBaseQuery({ baseUrl: "/ui-server/api" }),
```
1. No `/ui-server` prefix
2. Use `/api/data` here.
```suggestion
  baseQuery: fetchBaseQuery({ baseUrl: "/api/data" }),
```

andre-code replied:
sure, solved here [f147103](https://github.com/SwissDataScienceCenter/renku-ui/pull/3527/commits/f1471032d944adf9d773176b2550c2e6f4820dfe)

leafty commented on client/src/features/projects/projectMigration.api.ts:
```diff
@@ -0,0 +1,21 @@
+    getMigration: builder.query<Project, number>({
+      query: (projectId) => {
+        return {
+          url: `/data/renku_v1_projects/${projectId}/migrations`,
```
```suggestion
          url: `renku_v1_projects/${projectId}/migrations`,
```

andre-code replied:
sure, solved here [f147103](https://github.com/SwissDataScienceCenter/renku-ui/pull/3527/commits/f1471032d944adf9d773176b2550c2e6f4820dfe)

leafty commented on client/src/features/projects/projectMigration.api.ts:
Prefer using code generation instead of hand-writing client-side API handlers.

leafty commented on client/src/features/project/utils/projectMigration.utils.ts:
This file has no need to exist!
1. Use utilities from `react-router`: `import { generatePath } from "react-router-dom-v5-compat";`
2. Use the `ABSOLUTE_ROUTES` constant for routes.

andre-code replied:
you are right, done here sure [f147103](https://github.com/SwissDataScienceCenter/renku-ui/pull/3527/commits/f1471032d944adf9d773176b2550c2e6f4820dfe)

leafty commented on client/src/styles/bootstrap_ext/_button.scss:
Global style changes need a rationale.

andre-code replied:
fixed here [f147103](https://github.com/SwissDataScienceCenter/renku-ui/pull/3527/commits/f1471032d944adf9d773176b2550c2e6f4820dfe)

leafty commented on client/src/styles/components/_jupyter_notebook.scss:
Global style changes need a rationale. Though, in this case it is nice to remove styling that spills all over the UI.

andre-code replied:
fixed here [f147103](https://github.com/SwissDataScienceCenter/renku-ui/pull/3527/commits/f1471032d944adf9d773176b2550c2e6f4820dfe)

leafty commented on client/src/styles/components/_renku_form.scss:
Global style changes need a rationale.

leafty reviewed (commented):
Thank you @andre-code 👏 

Some first comments below.

leafty commented on client/src/features/project/components/projectMigration/ProjectEntityMigration.tsx:
```diff
@@ -0,0 +1,391 @@
+  const onSubmit = useCallback(
+    (data: ProjectMigrationForm) => {
+      if (!containerImage) return;
+      const nowFormatted = DateTime.now().toFormat("yyyy-MM-dd HH:mm:ss");
```
Suggestion:
```suggestion
      const nowFormatted = toHumanDateTime({ datetime: DateTime.now() });
```
This will ensure that the datetime format is consistent.

leafty commented on client/src/features/project/components/projectMigration/ProjectEntityMigration.tsx:
```diff
@@ -0,0 +1,391 @@
+      migrateProject({
+        projectMigrationPost: dataMigration,
+        v1Id: parseInt(projectMetadata.id),
+      } as PostRenkuV1ProjectsByV1IdMigrationsApiArg);
```
Question: why is this typing required?

leafty commented on client/src/features/project/components/projectMigration/ProjectEntityMigration.tsx:
```diff
@@ -0,0 +1,391 @@
+          )}
+          {!containerImage && !isFetchingData && (
+            <ErrorAlert dismissible={false}>
+              Container image not available, it is building or not exist
```
```suggestion
              Container image not available, it does not exist or is currently building.
```

leafty commented on client/src/features/project/components/projectMigration/ProjectMigrationForm.tsx:
```diff
@@ -0,0 +1,128 @@
+      shouldValidate: true,
+    });
+  }, [setValue, currentName]);
+  const url = `renkulab.io/v2/projects/${currentNamespace ?? "<Owner>"}/`;
```
See: https://github.com/SwissDataScienceCenter/renku-ui/issues/3517

andre-code replied:
fixed!

andre-code commented on client/src/features/project/components/projectMigration/ProjectEntityMigration.tsx:
```diff
@@ -0,0 +1,391 @@
+      : "";
+  }, [projectMigration]);
+
+  const projectMetadata: unknown = useLegacySelector<string>(
```
No needed, removed.

andre-code commented on client/src/features/project/components/projectMigration/ProjectEntityMigration.tsx:
```diff
@@ -0,0 +1,391 @@
+        toggle={toggle}
+        description={description?.isLoading ? undefined : description?.value}
+        tagList={tagList}
+        projectMetadata={projectMetadata as ProjectMetadata}
```
Not needed, removed.

andre-code commented:
> Partial review below.
> 
> This text does not make sense to me. Workflows have nothing to do with the `git lfs` command. `git lfs` was used to pull files in datasets for Renku 1.0. ![Screenshot 2025-03-07 at 09 16 10](https://private-user-images.githubusercontent.com/951086/420258759-5def6e3a-cf2c-4e49-82bb-ac8210381a8a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE3NzYwNjMsIm5iZiI6MTc0MTc3NTc2MywicGF0aCI6Ii85NTEwODYvNDIwMjU4NzU5LTVkZWY2ZTNhLWNmMmMtNGU0OS04MmJiLWFjODIxMDM4MWE4YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxMlQxMDM2MDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ODM2ZDM1MWQ3NjQyMjdlZGIyYjFkMzdhZTQ5OTdiYWJiYzRlMjQ2ODQ3YzJlN2YyYTEwMGI1YmYwNjdiYjNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.t2eNTJ77dG510ke6jIc_X2NSi-RilU-sw6YtsqBQXt8)

Thanks @leafty for your review!

I’ve fixed the copy for Workflows—let me know if that version makes more sense.
Regarding the command, args, port... parameters for migrating a project, I’ve moved those to the backend, so they are no longer needed here.

leafty reviewed (dismissed):
LGTM :clap: Great!

leafty reviewed (approved):
LGTM
